### PR TITLE
FEATURE: Master-Slave Redis configuration with fallback and switch over.

### PR DESCRIPTION
**Pull request by tgxworld**

### CODE REVIEW

**Comment by tgxworld** [comment]
@SamSaffron  Please review. Will need to publish a new MessageBus version :)

File: Gemfile [hunk 1]
```diff
@@ -38,7 +38,8 @@ gem 'mail'
 gem 'mime-types', require: 'mime/types/columnar'
 
 gem 'hiredis'
-gem 'redis', require:  ["redis", "redis/connection/hiredis"]
+# https://github.com/redis/redis-rb/pull/591
+gem 'redis', require:  ["redis", "redis/connection/hiredis"], github: 'tgxworld/redis-rb'
```
Comment (SamSaffron) [inline Gemfile]: hmmm very very uneasy with this, is the a PR for it? also don't source from github publish a gem instead

Comment (SamSaffron) [inline Gemfile]: let's freedom patch this the test will cover that it is still working on redis upgrades. we just need to remember to remove the freedom patch so add a note there.

File: spec/components/discourse_redis_spec.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  describe DiscourseRedis::FallbackHandler do
+    it 'should switch back to the master server' do
+      begin
+        fallback_handler.master = false
+        $redis.without_namespace.expects(:get).raises(Redis::CommandError.new("READONLY"))
+        Redis::Client.any_instance.expects(:call).with([:info]).returns(DiscourseRedis::FallbackHandler::MASTER_LINK_STATUS)
+        Redis::Client.any_instance.expects(:call).with([:client, [:kill, 'type', 'normal']])
+
+        current_threads = Thread.list
+        expect { $redis.get('test') }.to change { Thread.list.count }.by(1)
```
Comment (SamSaffron) [inline spec/components/discourse_redis_spec.rb]: this is not really a good test ... its testing for a side effect of a very particular implementation. this really should not matter.

File: spec/components/discourse_redis_spec.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        (Thread.list - current_threads).each(&:join)
```
Comment (SamSaffron) [inline spec/components/discourse_redis_spec.rb]: this is very weird...

File: spec/components/discourse_redis_spec.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        expect(fallback_handler.master).to eq(true)
```
Comment (SamSaffron) [inline spec/components/discourse_redis_spec.rb]: this is fine

File: spec/components/discourse_redis_spec.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        expect(Discourse.recently_readonly?).to eq(false)
```
Comment (SamSaffron) [inline spec/components/discourse_redis_spec.rb]: this is fine

File: spec/components/discourse_redis_spec.rb (hunk 3 above)
Comment (tgxworld) [inline spec/components/discourse_redis_spec.rb]: hmm why so? We need to wait for the thread to complete otherwise we'll get randomly failing tests.

Comment (tgxworld) [inline spec/components/discourse_redis_spec.rb]: @SamSaffron I updated the PR per your comments. Just not sure about this one.

**Comment by tgxworld** [comment]
@SamSaffron I'm merging this. Let me know if the specs is ok :)

**Pull request merged**